### PR TITLE
Query Builder for mysql [knex.js]

### PR DESCRIPTION
**Pull request by jyotirmay123**
Query builder from NLP output to fetch mysql data.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -11,8 +11,3 @@ install:
 script:
   - npm run test:server
   - npm run test:ui
-
-deploy:
-  provider: azure_web_apps
-  on: develop
-  verbose: true
```
Comment (csehlke) [inline .travis.yml:18]: is there a particular reason we removed this?

**Review by csehlke — APPROVED** [review]

File: server/utils/nlptodata.js [hunk 2]
```diff
@@ -4,93 +4,123 @@ var stringSimilarity = require('string-similarity');
...
-    data = {};
-    for (var column in arrayOfColumns) {
-        var name = arrayOfColumns[column].name;
-        data[name] = [];
-        var series = arrayOfColumns[column].series;
-        series.forEach(function (value) {
-            data[name].push(value);
+                fulfill(data);
+            } else {
+                reject("Sorry, your query is meaningless to us.");
```
Comment (csehlke) [inline server/utils/nlptodata.js:104]: sounds harsh, but its fine 😄

File: UI/src/app/home/home.component.ts [hunk 3]
```diff
@@ -14,7 +14,7 @@ export class HomeComponent implements AfterViewInit, OnInit {
     @ViewChild("textInput") textInput: TextInputComponent;
     @ViewChild("chart") chart: ChartComponent;
 
-    data: Trace[];
+    data: Trace;
```
Comment (csehlke) [inline UI/src/app/home/home.component.ts:17]: why did you remove the Array? we could display more than one trace, so we need the array of traces? 🤔

File: server/utils/nlptodata.js (hunk 2 above)
Comment (jyotirmay123) [reply server/utils/nlptodata.js:104]: haaaa haa..  You can change it to something softer.

File: UI/src/app/home/home.component.ts [hunk 4]
```diff
@@ -58,13 +58,23 @@ export class HomeComponent implements AfterViewInit, OnInit {
...
+            console.log(this.chart);
+            this.chart.plot();
+            let promise = this.backendConnector.requestData(this.textInput.getTextInput(), this.dataset.name, this);
+            promise.then(function(err){
+                console.log(err);
+                throw err;
+            }, function(that){
+                console.log(that.chart);
+                that.chart.plot();
+            });
```
Comment (csehlke) [inline UI/src/app/home/home.component.ts:70]: this should be moved to the backend-connector service

File: .travis.yml (hunk 1 above)
Comment (jyotirmay123) [reply .travis.yml:18]: It says deployment is not done in travis run every-time. We are good in running test cases but deployment is not configured properly yet with our ever changing scenarios. So removed it, as without clearing this step I cannot merge where as that is not configured properly.

File: UI/src/app/services/backend-connector.service.ts [hunk 5]
```diff
@@ -26,27 +27,22 @@ export class BackendConnectorService {
...
-                if(data) {
-                    const val = Data.parseData(data as JsonData);
-                    this.parseData(val.traces);
-                    this.parseLayout(val.layout);
-                    this.parseOptions(val.options);
+            .get('http://localhost:3001/api/nlptodata', 
+            {
+                params: {
+                    userquery: 'Plot number of employees for each sex as a pie chart for each departments',
+                     dataset: 'core_data'
```
Comment (csehlke) [inline UI/src/app/services/backend-connector.service.ts:36]: You want the name of the data set as a string here right?

File: UI/src/app/text-input/text-input.component.ts [hunk 6]
```diff
@@ -7,10 +7,12 @@ import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
 })
 export class TextInputComponent implements OnInit {
 
-    textInput: string;
+    textInput: string = 'Plot a pie chart for number of employees in each department';
```
Comment (csehlke) [inline UI/src/app/text-input/text-input.component.ts:10]: Why do you assign that string twice?

File: UI/src/app/home/home.component.ts (hunk 4 above)
Comment (jyotirmay123) [reply UI/src/app/home/home.component.ts:70]: yeah.. I was just trying but not able to.

File: UI/src/app/text-input/text-input.component.ts (hunk 6 above)
Comment (jyotirmay123) [reply UI/src/app/text-input/text-input.component.ts:10]: no no..  This needs to changed back to textInput: string;

File: UI/src/app/services/backend-connector.service.ts (hunk 5 above)
Comment (jyotirmay123) [reply UI/src/app/services/backend-connector.service.ts:36]: yes, I changed it here.

File: UI/src/app/home/home.component.ts (hunk 3 above)
Comment (jyotirmay123) [reply UI/src/app/home/home.component.ts:17]: Yes, It also need to change back to trace[];

File: UI/src/app/home/home.component.ts [hunk 7]
```diff
@@ -58,13 +58,23 @@ export class HomeComponent implements AfterViewInit, OnInit {
             console.log("Plot was pressed");
             console.log(this.textInput.getTextInput());
             console.log(this.dataset.name);
-            this.backendConnector.requestData(this.textInput.getTextInput(), this.dataset.id);
-            setTimeout( () => {
+            console.log(this.chart);
+            this.chart.plot();
+            let promise = this.backendConnector.requestData(this.textInput.getTextInput(), this.dataset.name, this);
```
Comment (csehlke) [inline UI/src/app/home/home.component.ts:63]: what is the last atribute `this` for?

File: UI/src/app/home/home.component.ts [hunk 8]
```diff
@@ -58,13 +58,23 @@ export class HomeComponent implements AfterViewInit, OnInit {
...
+            promise.then(function(err){
+                console.log(err);
+                throw err;
+            }, function(that){
+                console.log(that.chart);
+                that.chart.plot();
```
Comment (csehlke) [inline UI/src/app/home/home.component.ts:69]: shouldnt they be the other way around?

the first attribute of `then`is called `onfulfilled` and the secound is called `onrejected`

File: UI/src/app/services/backend-connector.service.ts [hunk 9]
```diff
@@ -11,10 +12,10 @@ export class BackendConnectorService {
     private readonly connectionTimeout = 100000;
 
 
-    private _data: BehaviorSubject<Map<number, Trace[]>> = new BehaviorSubject<Map<number, Trace[]>>(new Map());
+    private _data: BehaviorSubject<Map<number, Trace>> = new BehaviorSubject<Map<number, Trace>>(new Map());
```
Comment (csehlke) [inline UI/src/app/services/backend-connector.service.ts:15]: missing [ ]
I know it is fixed on the other branch, but I ll note it down so I don't forget to check again

File: UI/src/app/services/backend-connector.service.ts [hunk 10]
```diff
@@ -11,10 +12,10 @@ export class BackendConnectorService {
...
     private _options: BehaviorSubject<Map<number, Options>> = new BehaviorSubject<Map<number, Options>>(new Map());
     private _layout: BehaviorSubject<Map<number, Layout>> = new BehaviorSubject<Map<number, Layout>>(new Map());
-    public data: Observable<Map<number, Trace[]>> = this._data.asObservable();
+    public data: Observable<Map<number, Trace>> = this._data.asObservable();
```
Comment (csehlke) [inline UI/src/app/services/backend-connector.service.ts:18]: missing [ ]

File: UI/src/app/services/backend-connector.service.ts [hunk 11]
```diff
@@ -1,8 +1,9 @@
 import { Injectable } from '@angular/core';
-import { HttpClient } from '@angular/common/http';
+import { HttpClient, HttpParams } from '@angular/common/http';
 import { ScatterData } from 'plotly.js/lib/core';
 import { Data, JsonData, Layout, Options, Trace } from "../../../models";
```
Comment (csehlke) [inline UI/src/app/services/backend-connector.service.ts:4]: we could clean up the unused imports a bit

File: UI/src/app/services/backend-connector.service.ts [hunk 12]
```diff
@@ -63,20 +59,126 @@ export class BackendConnectorService {
...
-        return this.http
-            .post('localhost:3001/api', JSON.stringify({userinput: userInput, datasetid: datasetId}))
-            .retry(this.connectionRetries)
-            .timeout(this.connectionTimeout)
-            .map((response: Response) =>
-                response.json()
-            )
+    public requestData(userInput: string, datasetName: string, that){
+        return new Promise(function(reject, fullfill){
+            this.that = that;
```
Comment (csehlke) [inline UI/src/app/services/backend-connector.service.ts:64]: what is `that` ? Could you add a type to it?

File: UI/src/app/services/backend-connector.service.ts [hunk 13]
```diff
@@ -26,27 +27,22 @@ export class BackendConnectorService {
...
-            .map((response: Response) =>
-                response.json() || {}
-            )
-            // .catch(this.handleError)
-            .subscribe(data => {
-                if(data) {
-                    const val = Data.parseData(data as JsonData);
-                    this.parseData(val.traces);
-                    this.parseLayout(val.layout);
-                    this.parseOptions(val.options);
```
Comment (csehlke) [inline UI/src/app/services/backend-connector.service.ts:42]: you are missing the parsing part, otherwise it does not get saved to the Observables

File: UI/src/app/services/backend-connector.service.ts [hunk 14]
```diff
@@ -63,20 +59,126 @@ export class BackendConnectorService {
...
+                params: {
+                    userquery: userInput,
+                     dataset: datasetName
+                }
+            })
             .subscribe((data) => {
-                console.log(data);
+                console.log(this.that);
+                var t:Trace = new Trace(data["columnA"], data["columnB"], data["plotType"])
+                this.parseData(t)
```
Comment (csehlke) [inline UI/src/app/services/backend-connector.service.ts:75]: oh I see, I divided send and request, but we should do it as one, thats correct, maybe we should add the deleted lines from the getData methode in here

File: UI/src/app/text-input/text-input.component.ts [hunk 15]
```diff
@@ -7,10 +7,12 @@ import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
...
     @Output() plot = new EventEmitter();
 
-    constructor() {}
+    constructor() {
+        this.textInput = 'Plot a pie chart for number of employees in each department';
```
Comment (csehlke) [inline UI/src/app/text-input/text-input.component.ts:14]: Its not working to set the default text in here since it is done in the html file to it.
